src/error.rs

```diff
@@ -58,6 +58,37 @@ pub enum Error {
 
     Utf8Error(Utf8Error),
     TrailingCharacters,
+
+    ExpectedDifferentType {
+        expected: String,
+        found: UnexpectedSerdeTypeValue,
+    },
+    InvalidValueForType {
+        expected: String,
+        found: UnexpectedSerdeTypeValue,
+    },
```

Comment:
What's the difference between those two errors? In fact, I can't understand at all what the "DifferentType" means. (:
When deserializer expects an integer, but I provide a string, which error is raised?

Reply:
The only difference is that they come from different error methods in `serde`: [`invalid_type`](https://docs.rs/serde/latest/serde/de/trait.Error.html#method.invalid_type) and [`invalid_value`](https://docs.rs/serde/latest/serde/de/trait.Error.html#method.invalid_value). When your field is an integer but you give a string, neither will be called since `ron` handles that one natively. It is actually quite difficult to observe that one in the wild. The `invalid_value` error is more common, e.g. when you try to deserialise `0` into `NonZero`.

Reply:
Perhaps it would be useful to go over the existing errors and reclassify them as type or value errors as well so there's no overlap - this could also allow us to clean up those two errors from `serde` based on what we think they should be